### PR TITLE
Add vocab masks to Whisper export with beam search

### PR DESCRIPTION
**Pull request by kunal-vaishnavi**
### Description
This PR adds flags for exporting Whisper with vocab masks for logits processing. This PR also sets `input_features` back to FP32 precision for the user and casts `input_features` to FP16 precision when needed.

### Motivation and Context
This helps enable specific logits processing for the exported Whisper model.

### CODE REVIEW

File: onnxruntime/python/tools/transformers/models/whisper/whisper_chain.py [hunk 1]
```diff
@@ -44,6 +44,15 @@ def chain_model(args):
         beam_inputs.append("logits_processor")
     beam_outputs = ["sequences"]
 
+    if args.precision == Precision.FLOAT16:
+        cast_node = helper.make_node(
+            "Cast",
+            inputs=["input_features"],
+            outputs=["input_features_fp16"],
+            name="CastToFp16",
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/models/whisper/whisper_chain.py:52]: nit: more specific like CastInputFeaturesToFp16.

File: onnxruntime/python/tools/transformers/models/whisper/whisper_chain.py [hunk 2]
```diff
@@ -115,7 +129,8 @@ def chain_model(args):
 
     opset_import = [helper.make_opsetid(domain="com.microsoft", version=1), helper.make_opsetid(domain="", version=17)]
 
-    beam_graph = helper.make_graph([node], "beam-search-test", graph_inputs, graph_outputs, initializers)
+    graph_nodes = ([cast_node] if args.precision == Precision.FLOAT16 else []) + [node]
+    beam_graph = helper.make_graph(graph_nodes, "beam-search-test", graph_inputs, graph_outputs, initializers)
     beam_model = helper.make_model(beam_graph, producer_name="pytorch", opset_imports=opset_import)
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/models/whisper/whisper_chain.py:134]: change the producer to onnxruntime.

File: onnxruntime/python/tools/transformers/models/whisper/whisper_chain.py (hunk 1 above)
Comment (zhanghuanrong) [reply onnxruntime/python/tools/transformers/models/whisper/whisper_chain.py:52]: length_penalty, repetition_penalty also need cast

**Review by petermcaughan — APPROVED** [review]
